src/core-manager/core-replication-state.js

```diff
@@ -0,0 +1,395 @@
+ * @property {number} coreLength known (sparse) length of the core
+ * @property {PeerSimpleState} localState local state
+ * @property {Record<PeerId, RemotePeerSimpleState>} remoteStates map of state of all known peers
+ */
+/**
+ * @typedef {object} CoreReplicationEvents
+ * @property {() => void} update
+ */
+
+/**
+ * Track replication state for a core identified by `discoveryId`. Can start
+ * tracking state before the core instance exists locally, via the "preHave"
+ * messages received over the project creator core.
+ *
+ * Because deriving the state is expensive (it iterates through the bitfields of
+ * all peers), this is designed to be pull-based: an `update` event signals that
+ * the state is updated, but does not pass the state. The consumer can "pull"
+ * the state when it wants it via `coreReplicationState.getState()`.
+ *
+ * Each peer (including the local peer) has a state of:
+ *   1. `have` - number of blocks the peer has locally
+ *   2. `want` - number of blocks the peer wants, and at least one peer has
+ *   3. `wanted` - number of blocks the peer has that at least one peer wants
+ *   4. `missing` - number of blocks the peer wants but no peer has
+ *
+ * @extends {TypedEmitter<CoreReplicationEvents>}
+ */
+export class CoreReplicationState extends TypedEmitter {
+  /** @type {import('hypercore')<'binary', Buffer>} */
+  #core
+  /** @type {InternalState['remoteStates']} */
+  #remoteStates = new Map()
+  /** @type {InternalState['localState']} */
+  #localState = new PeerState()
+  #discoveryId
+  /** @type {DerivedState | null} */
+  #cachedState = null
+
+  /**
+   * @param {string} discoveryId Discovery ID for the core that this is representing
+   */
+  constructor(discoveryId) {
+    super()
+    this.#discoveryId = discoveryId
+  }
+
+  /** @type {() => DerivedState} */
+  getState() {
+    if (this.#cachedState) return this.#cachedState
+    return deriveState({
+      length: this.#core?.length,
+      localState: this.#localState,
+      remoteStates: this.#remoteStates,
+    })
+  }
+
+  /**
+   * Called whenever the state changes, so we clear the cache because next call
+   * to getState() will need to re-derive the state
+   */
+  #update() {
+    this.#cachedState = null
+    this.emit('update')
+  }
+
+  /**
+   * Attach a core. The replication state can be initialized without a core
+   * instance, because we could receive peer want and have states via extension
+   * messages before we have the core key that allows us to create a core
+   * instance.
+   *
+   * @param {import('hypercore')<'binary', Buffer>} core
+   */
+  attachCore(core) {
+    // @ts-ignore - we know discoveryKey exists here
+    const discoveryId = keyToId(core.discoveryKey)
+    if (discoveryId !== this.#discoveryId) {
+      throw new Error('discoveryId does not match')
+    }
+    if (!this.#core) return
+
+    this.#core = core
+    // @ts-ignore - internal property
+    this.#localState.setHavesBitfield(core?.core?.bitfield)
```

Comment:
Small suggestion since I've been bit by this before, but mind changing this placement of the ignore on the intended call?

```suggestion
    this.#localState.setHavesBitfield(
      // @ts-ignore - internal property
      core?.core?.bitfield
    )
```

Reply:
very good idea - I've been bitten by this before too